File: app/src/main/java/de/sneak/sneakpeek/ui/fragment/StudiosFragment.java
package de.sneak.sneakpeek.ui.fragment;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import java.util.List;

import de.sneak.sneakpeek.R;
import de.sneak.sneakpeek.adapter.StudiosAdapter;
import de.sneak.sneakpeek.data.Score11Movie;
import de.sneak.sneakpeek.service.MovieRepository;
import rx.Subscription;
import rx.functions.Action1;
import rx.subscriptions.CompositeSubscription;

public class StudiosFragment extends Fragment {

    private static final String TAG = MovieFragment.class.getSimpleName();

    private CompositeSubscription subscriptions;
    private SwipeRefreshLayout swipeRefreshLayout;
    private StudiosAdapter studiosAdapter;

    public static StudiosFragment newInstance() {
        return new StudiosFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        swipeRefreshLayout = (SwipeRefreshLayout) inflater.inflate(R.layout.fragment_studios, container, false);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                setMovies();
            }
        });

        RecyclerView recyclerView = (RecyclerView) swipeRefreshLayout.findViewById(R.id.fragment_studios_recycler_view);

        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        studiosAdapter = new StudiosAdapter();
        recyclerView.setAdapter(studiosAdapter);

        subscriptions = new CompositeSubscription();

        return swipeRefreshLayout;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setMovies();
    }

    @Override
    public void onStop() {
        super.onStop();
        subscriptions.unsubscribe();
    }

    public void setMovies() {
        Subscription subscription = MovieRepository.getInstance().fetchStudios()
                .subscribe(new Action1<List<Score11Movie>>() {
                    @Override
                    public void call(List<Score11Movie> movieRepository) {
                        studiosAdapter.addAll(movieRepository);
                        swipeRefreshLayout.setRefreshing(false);
                    }
                }, new Action1<Throwable>() {
                    @Override
                    public void call(Throwable throwable) {
                        swipeRefreshLayout.setRefreshing(false);
                        Toast.makeText(getContext(), "Failed to fetch movies", Toast.LENGTH_SHORT).show();
                        Log.e(TAG, "Failed to fetch movie predictions", throwable);
                    }
                });

        subscriptions.add(subscription);
    }
}

File: app/src/main/java/de/sneakpeek/ui/main/fragment/StudiosFragment.kt
package de.sneakpeek.ui.main.fragment

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import de.sneakpeek.R
import de.sneakpeek.adapter.StudiosAdapter
import de.sneakpeek.service.MovieRepository
import de.sneakpeek.util.inflate
import de.sneakpeek.view.FastScroll
import de.sneakpeek.view.FastScrollItemDecorator
import io.reactivex.disposables.CompositeDisposable

class StudiosFragment : Fragment() {

    private var subscriptions: CompositeDisposable = CompositeDisposable()
    private val studiosAdapter: StudiosAdapter by lazy { StudiosAdapter(emptyList()) }
    var fastScroll: FastScroll? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        fastScroll = container?.inflate(R.layout.fragment_studios) as FastScroll

        fastScroll?.let {
            it.setHasFixedSize(true)
            it.layoutManager = LinearLayoutManager(context)
            it.adapter = studiosAdapter

            it.addItemDecoration(FastScrollItemDecorator(context))
            it.itemAnimator = DefaultItemAnimator()
        }

        return fastScroll
    }

    override fun onStart() {
        super.onStart()
        subscriptions = CompositeDisposable()
        loadStudios()
    }

    override fun onStop() {
        super.onStop()
        subscriptions.dispose()
    }

    fun loadStudios() {
        val subscription = MovieRepository(context).getStudios()
                ?.subscribe({
                    studiosAdapter.addAll(it)
                    fastScroll?.forceReSetup()
                }) { throwable ->
                    Toast.makeText(context, "Failed to fetch movies", Toast.LENGTH_SHORT).show()
                    Log.e(TAG, "Failed to fetch movie moviePredictions", throwable)
                }

        subscriptions.add(subscription)
    }

    companion object {

        private val TAG = StudiosFragment::class.java.simpleName

        fun newInstance(): StudiosFragment {
            return StudiosFragment()
        }
    }
}